projects/pgai/pgai/vectorizer/vectorizer.py

```diff
@@ -622,6 +789,67 @@ async def _embed_and_write(self, conn: AsyncConnection, items: list[SourceRow]):
 
         return len(records)
 
+    @tracer.wrap()
+    async def _embed_and_write_from_batch(
+            self,
+            conn: AsyncConnection,
+            batch: openai.types.Batch,
+            client: OpenAI,
+    ):
+        """
+        Embeds the items and writes them to the database.
+
+        - Deletes existing embeddings for the items.
+        - Generates the documents to be embedded, chunks them, and formats the chunks.
+        - Sends the documents to the embedding provider and writes embeddings
+          to the database.
+        - Logs any non-fatal errors encountered during embedding.
+
+        Args:
+            conn (AsyncConnection): The database connection.
+            batch: The batch as retrieved from OpenAI's api.
+            client: The OpenAI client to use.
+
+        Returns:
+            int: The number of records written to the database.
+        """
+        batch_file = client.files.content(batch.output_file_id)
+
+        batch_data = batch_file.text.strip().split('\n')
+        num_records = 0
+        all_items = []
+        all_records: list[EmbeddingRecord] = []
+
+        # Fetch all chunks from ai.embedding_batch_chunks where the embedding_batch_id is batch.id
+        async with conn.cursor() as cursor:
+            await cursor.execute(
+                "SELECT id, text FROM ai.embedding_batch_chunks WHERE embedding_batch_id = %s",
+                (batch.id,)
+            )
+            embedding_batch_chunks = {row[0]: row[1] for row in await cursor.fetchall()}
+            
+            for line in batch_data:
+                json_line = json.loads(line)
+                if "custom_id" in json_line and "response" in json_line:
+
+                    custom_id = json_line['custom_id']
+                    pk_names, document_id, chunk_seq = custom_id.split(':::')
+                    embedding_data = json_line['response']['body']['data'][0]['embedding']
+
+                    resolved_id = document_id.split(',')
+                    resolved_pk = pk_names.split(',')
+                    item = {pk: id_value for pk, id_value in zip(resolved_pk, resolved_id)}
+                    item[self.vectorizer.config.chunking.chunk_column] = embedding_batch_chunks[custom_id]
+
+                    all_items.append(item)
+                    all_records.append([resolved_id + [chunk_seq, embedding_batch_chunks[custom_id]] + [np.array(embedding_data)]])
+
+        await self._delete_embeddings(conn, all_items)
+        for records in all_records:
+            await self._copy_embeddings(conn, records)
```

Comment:
This fails with postgres running out of memory:

```
ERROR:  out of memory
DETAIL:  Cannot enlarge string buffer containing 0 bytes by 2130706440 more bytes.
CONTEXT:  COPY wiki_orig_embeddings_store, line 1, column wiki_id
STATEMENT:  COPY "public"."wiki_orig_embeddings_store" ("wiki_id" ,"paragraph_id", chunk_seq, chunk, embedding) FROM STDIN WITH (FORMAT BINARY)
```

It first failed when I tried inserting all records at once, but inserting them one by one does not seem to work either.